src/vsr/replica.zig

```diff
@@ -230,28 +242,50 @@ pub fn ReplicaType(
         /// The number of ticks before a primary or backup broadcasts a ping to other replicas.
         /// TODO Explain why we need this (MessageBus handshaking, leapfrogging faulty replicas,
         /// deciding whether starting a view change would be detrimental under some network partitions).
+        /// (status=normal replicas)
         ping_timeout: Timeout,
 
         /// The number of ticks without enough prepare_ok's before the primary resends a prepare.
+        /// (status=normal primary, pipeline has prepare with !ok_quorum_received)
         prepare_timeout: Timeout,
 
+        /// The number of ticks waiting for a prepare_ok.
+        /// When triggered, set primary_abdicating=true, which pauses outgoing commit heartbeats.
+        /// (status=normal primary, pipeline has prepare with !ok_quorum_received)
+        primary_abdicate_timeout: Timeout,
+
         /// The number of ticks before the primary sends a commit heartbeat:
         /// The primary always sends a commit heartbeat irrespective of when it last sent a prepare.
         /// This improves liveness when prepare messages cannot be replicated fully due to partitions.
-        commit_timeout: Timeout,
-
-        /// The number of ticks without hearing from the primary before starting a view change.
-        /// This transitions from .normal status to .view_change status.
-        normal_status_timeout: Timeout,
-
-        /// The number of ticks before a view change is timed out:
-        /// This transitions from `view_change` status to `view_change` status but for a newer view.
+        /// (status=normal primary)
+        commit_message_timeout: Timeout,
+
+        /// The number of ticks without a heartbeat.
+        /// Reset any time the backup receives a heartbeat from the primary.
+        /// Triggers SVC messages. If an SVC quorum is achieved, we will kick off a view-change.
+        /// (status=normal backup)
+        normal_heartbeat_timeout: Timeout,
+
+        /// The number of ticks before resetting the SVC quorum.
+        /// (status=normal|view-change)
+        start_view_change_window_timeout: Timeout,
```

Comment:
Hm, why do we need this timeout, and what's the overall flow here? If I understand the implementation correctly, this timeout always ticks, and just resets the quorum. This ... feels weird. 

In particular, the timeout here seems to be orthogonal to the process we use so start sending SVCs. This means we may accidentally trip this timeout just before we collected an SVC quorum? 

I would maybe expect us to just reset the quorum on commit?

Reply:
Suppose:
1. There is a temporary network issue
2. We collect a few SVCs
3. The network issue heals
4. (much time passes)
5. There is another network issue.
6. At this point we want to start collecting SVCs, but we want to start from scratch, not from wherever our old quorum left off.

Reply:
Hm, and `The network issue heals` means "we received a commit" I think? So that's the place to reset the quorum? I agree that we should reset the quorum at some point, I don't understand why we use timer for that.

Reply:
Except it may be the the _other_ replicas received a commit (so they stopped sending SVCs, they're healthy), but we ourselves didn't see a commit — all we know is we stopped seeing SVCs.

i.e. we are partitioned from the primary (not receiving `commit`s). Some other replicas were temporarily partitioned from the primary, but recovered. Their SVCs are stale and shouldn't count towards a quorum unless they are recent enough to reflect those other replica's actual (current) view of the network.

Reply:
Oh, I think I see what you mean -- we don't want SVCs to be too stale. 

Yeah, it semes that we need some timeout here, but perhaps we should start it after sending/receiving the first SVC, rather than when transitioning to normal?

The case I worry about is that everyone notices that the primary is dead, everyone sends an SVC, we receive SVCs and it just so happens that this timeout triggers, and we clear SVCs, though we received all of them like 1 ms ago.

Reply:
What about `self.start_view_change_window_timeout.reset();` any time we receive any SVC?

Reply:
I think that’s ok, although doesn’t yet feel like the perfect soln (to be fair, my proposal to start the timer on SVC reciting also feels adhoc)

Reply:
I think the "start timer on SVC" solution reflects the concept better, but I like keeping the invariants of "when does the timer run" from getting too complicated, even when that means putting an early-return into to timeout handler function.

Reply:
Maybe ideally each SVC would expire independently, but I don't want to manage that many timers.

Reply:
Thinking more, I don’t like “reset on any SVC”: if two replicas are partitioned together, they reset each other’s timers, and keep stale SVCs from other replicas.

> Maybe ideally each SVC would expire independently, but I don't want to manage that many timers.

Yeah, that would be inelegant I would think!

> better, but I like keeping the invariants of "when does the timer run" from getting too complicated,

Uhu! At the same time, the current invariant is “timer always runs”, which has the drawback that it doesn’t give you any extra space for sanity checking or what not.

I think what I am suggesting is “timer is ticking if SVC quorum is non-empty”, which feels rather nice.

Reply:
Implemented, though instead of “timer is ticking if SVC quorum is non-empty” I went with "timer is ticking if SVC quorum has an SVC from someone _other than ourself_". The staleless/timeout isn't applicable to our own SVC, since we can keep that up-to-date more accurately based on the heartbeats.